### PR TITLE
Open sourcing the internal HParams plugin.

### PR DESCRIPTION
**Pull request by erzel**
* Detailed steps to verify changes work correctly (as executed by you)
- I've hooked it up to 'defaults_plugins' and saw that it works in open-source (though the current PR does not enable the plugin by default).
- I've executed 
  bazel test //tensorboard/...:all
  bazel run //tensorboard/plugins/hparams:web_components_tests and made sure the tests pass.                      

I am not sure about whether I'm importing tensorflow in the right way in the Python files in  //tensorboard/plugins/hparams so please make sure it's correct.

Thanks!
@nfelt @stephanwlee @manivaradarajan @wchargin PTAL.

### CODE REVIEW

File: tensorboard/plugins/hparams/tf_hparams_google_analytics_tracker/tf-hparams-google-analytics-tracker.html [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+  /**
+   * Create a new "ga" object the way it's prescribed by Google Analytics.
+   *
+   * Copied from Google Analytics, but styling applied to make the JS linter
+   * happy.
+   *
+   * @return {!function(...)} a "ga" object proxy ready for use.
+   */
+  function newGa() {
+    (function(i, s, o, g, r, a, m) {
```
Comment (wchargin) [inline tensorboard/plugins/hparams/tf_hparams_google_analytics_tracker/tf-hparams-google-analytics-tracker.html]: Using analytics _within Google_ is fine, but we want to be very clear
that we’re not tracking our open source users at all. From looking at
this change as a whole, I can see that this tracking code isn’t actually
used, but that’s not locally obvious at all.

I would be much more comfortable if we could remove this tracking code
from the open-source version of TensorBoard entirely. TensorBoard
already does this in `tensorboard/components/analytics.html`:

<https://github.com/tensorflow/tensorboard/blob/382df2a414524effab9099c95190efa4ccfdd06a/tensorboard/components/analytics.html#L19-L24>

For the Google-internal version of TensorBoard, this file is swapped out
for a version that actually uses Google Analytics; in the open source
version, it’s obvious from the code that no tracking is or can be
performed.

Could you please either use or emulate this mechanism?

Comment (erzel) [reply tensorboard/plugins/hparams/tf_hparams_google_analytics_tracker/tf-hparams-google-analytics-tracker.html]: Thanks Will. Sure. Let me do that and upload a new version.

**Comment by wchargin** [comment]
(Also, Travis is complaining that you’re missing LICENSE headers on all
your files, but that’s easy to fix:

```shell
./tensorboard/tools/license_test.sh | sed 1d |
xargs -d '\n' \
awk -i inplace \
    -v license="$(awk '! /^#/ { exit }; 1' tensorboard/lazy_test.py)" \
    'BEGINFILE { print license; }; 1'
```

)

**Comment by erzel** [comment]
Thanks @wchargin! PTAL.

**Comment by erzel** [comment]
Ok. I actually realized I've applied the license fix incorrectly to HTML files, and that the analytics-test is timing out. Please hold off reviewing. I'll fix that and let you know when I'm done.

**Comment by erzel** [comment]
@wchargin  ok. I hope I fixed all the reported issues now. PTAL

**Comment by wchargin** [comment]
> Ok. I actually realized I've applied the license fix incorrectly to
> HTML files

Oops, of course. my bad :-)

File: tensorboard/plugins/hparams/BUILD [hunk 2]
```diff
@@ -45,7 +52,74 @@ py_test(
...
+#    and put the test in a a file named: tf-hparams-<component>-test.html in
+#    that directory.
+# 2. Add a 'test_lib' tf_web_library rule to the BUILD file of the component
+#    that contains the test.
+# 3. Update the 'web_components_tests.html' file to load
+#    '../tf-hparams-<component>/test/tf-hparams-<component>-test.html"
+# 4. Add "//tensorboard/plugins/hparams/tf_hparams_<component>:test_lib"
+#    as a dependency to the rule below.
+#
+# TODO(erez): Simplify adding a new test. It should not require touching so many files.
```
Comment (stephanwlee) [inline tensorboard/plugins/hparams/BUILD]: Something like https://github.com/tensorflow/tensorboard/blob/1bfc270d81fa78730cf26d1bf1e98b1bf20a87ae/tensorboard/components/tf_backend/test/BUILD#L6-L14 should work

File: tensorboard/plugins/hparams/hparams_demo.py [hunk 3]
```diff
@@ -152,30 +179,31 @@ def run(logdir, session_id, hparams, group_name):
...
   # Now, augment the current temperature by this delta that we computed,
   # blocking the assignment on summary collection to avoid race conditions
   # and ensure that the summary always reports the pre-update value.
   with tf.control_dependencies([summ]):
     update_step = temperature.assign_add(delta)
 
-  sess = tf.compat.v1.Session()
-  sess.run(tf.compat.v1.global_variables_initializer())
-  for step in xrange(STEPS):
+  sess = tf.Session()
```
Comment (stephanwlee) [inline tensorboard/plugins/hparams/hparams_demo.py]: Please revert the change and use compat.v1. Also, please test your code against tf-nightly and tf-nightly-2.0-preview

File: tensorboard/plugins/hparams/BUILD (hunk 2 above)
Comment (erzel) [reply tensorboard/plugins/hparams/BUILD]: Hmm. So you now have web_testing in opensource. Cool. I just want to make sure these tests will run in google3 as well and it seems like the tf_web_test does nothing in google3, currently. Let me think about it some more. Meanwhile, can you continue to look over this PR and let me know if you have any other comments?

File: tensorboard/plugins/hparams/hparams_demo.py (hunk 3 above)
Comment (erzel) [reply tensorboard/plugins/hparams/hparams_demo.py]: Done.

**Comment by erzel** [comment]
Thanks @stephanwlee, PTAL!

File: tensorboard/plugins/hparams/hparams_demo.py [hunk 4]
```diff
@@ -31,8 +35,9 @@
 import shutil
 
 from six.moves import xrange  # pylint: disable=redefined-builtin
-import tensorflow as tf
-from tensorflow import flags
+import tensorflow.compat.v1 as tf
+from absl import flags
+from absl import app
```
Comment (stephanwlee) [inline tensorboard/plugins/hparams/hparams_demo.py]: please add absl dependencies on the BUILD file.

Comment (erzel) [reply tensorboard/plugins/hparams/hparams_demo.py]: Done.

File: tensorboard/plugins/hparams/hparams_demo.py [hunk 5]
```diff
@@ -20,6 +20,10 @@
 This demo is a slightly modified version of
 tensorboard/plugins/scalar/scalar_demo.py.
 """
+# TODO(erez): This code currently does not work with tf-nightly-2.0 since
+# it uses tf.compart.v1.summary.FileWriter which can't be used in eager
+# mode (which is the default in tensorflow V2). Fix this when we changet this
```
[the remark this answers is not in the document]
Comment (erzel) [reply tensorboard/plugins/hparams/hparams_demo.py]: Done.

**Comment by erzel** [comment]
Thanks @stephanwlee! PTAL.

File: tensorboard/plugins/hparams/backend_context_test.py [hunk 6]
```diff
@@ -20,7 +20,7 @@
 
 import operator
 
-import tensorflow as tf
+import tensorflow.compat.v1 as tf
```
[the remark this answers is not in the document]
Comment (erzel) [reply tensorboard/plugins/hparams/backend_context_test.py]: Done.

File: tensorboard/plugins/hparams/hparams_util.py [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+"""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import getpass
+import time
+
+from absl import app
```
[the remark this answers is not in the document]
Comment (erzel) [reply tensorboard/plugins/hparams/hparams_util.py]: I've added "//tensorboard:expect_absl_app_installed" which seems to be in line with
the dependency of "//tensorboard::expect_absl_logging_installed" of the  //tensorboard:loader target. Do you want me to do something else?

File: tensorboard/plugins/hparams/hparams_util_test.sh [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    values: { string_value:"1e-3" }
+    values: { string_value:"1e-4" }
+  }
+}
+hparam_infos: {
+  name: "num_keypoints"
+  display_name: "num_keypoints"
+  type: DATA_TYPE_STRING
+  domain_discrete: {
+     values: {string_value:"100" }
```
[the remark this answers is not in the document]
Comment (erzel) [reply tensorboard/plugins/hparams/hparams_util_test.sh]: Done.

**Comment by erzel** [comment]
Thanks @stephanwlee! PTAL.

File: tensorboard/plugins/hparams/tf_hparams_parallel_coords_view/tf-hparams-parallel-coords-view.html [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+        // See the property description in tf-hparams-query-pane.html
+        experimentName: String,
+        // See the property description in tf-hparams-query-pane.html
+        schema: Object,
+        // See the property description in tf-hparams-query-pane.html
+        sessionGroups: Array,
+      },
+
+      _closestOrSelected: function(closestSessionGroup, selectedSessionGroup) {
+        if (closestSessionGroup !== null) {
```
Comment (stephanwlee) [inline tensorboard/plugins/hparams/tf_hparams_parallel_coords_view/tf-hparams-parallel-coords-view.html]: `return closestSessionGroup || selectedSessionGroup`? 🙈

File: tensorboard/plugins/hparams/tf_hparams_scale_and_color_controls/tf-hparams-scale-and-color-controls.html [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+      _schemaChanged() {
+        const _this = this;
+        // Populate options.columns with a linear scale for each column (
+        // hparam or metric).
+        const schema = this.schema;
+        const newHParamColumn = (info, index) => {
+          return {
+            name: tf.hparams.utils.hparamName(info),
+            index: index,
+            scale: _this._isNumericColumn(index) ? "LINEAR" : "NON_NUMERIC",
```
Comment (stephanwlee) [inline tensorboard/plugins/hparams/tf_hparams_scale_and_color_controls/tf-hparams-scale-and-color-controls.html]: remove L191 and use `this`. You are using an arrow func.

File: tensorboard/plugins/hparams/tf_hparams_session_group_details/tf-hparams-session-group-details.html [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+          type: Boolean,
+          value: false
+        },
+
+        _requestData: {
+          type: Function,
+          value: function() {
+            return ({tag, run}) => new Promise((resolve, reject) => {
+              const url = this.apiUrl + "/metric_evals";
+              const xhr = new XMLHttpRequest();
```
Comment (stephanwlee) [inline tensorboard/plugins/hparams/tf_hparams_session_group_details/tf-hparams-session-group-details.html]: Please consider using the request manager (with API changes if required to support POST).

Also using `fetch` browser API would clean this up a lot.

File: tensorboard/plugins/hparams/tf_hparams_session_group_details/tf-hparams-session-group-details.html [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+      },
+
+      // Returns an array of the tensorboard 'runs' and 'tags' corresponding to
+      // the given metric in each of the sessions in the given session group.
+      _computeSeriesForSessionGroupMetric(sessionGroup, metricInfo) {
+        if (sessionGroup === null || metricInfo === null) {
+          return [];
+        }
+        let result = [];
+        sessionGroup.sessions.forEach(session => {
```
Comment (stephanwlee) [inline tensorboard/plugins/hparams/tf_hparams_session_group_details/tf-hparams-session-group-details.html]: consider using `filter` and `map`.

File: third_party/python.bzl [hunk 13]
```diff
@@ -44,6 +44,16 @@ def tensorboard_python_workspace():
         build_file = str(Label("//third_party:urllib3.BUILD")),
     )
 
+    http_archive(
+        name = "org_pythonhosted_attrs",
```
Comment (stephanwlee) [inline third_party/python.bzl]: [AI REQUIRED] Is this going to be the required dependency on TB? If so, add a line here
https://github.com/tensorflow/tensorboard/blob/6359c4492e8844ace5761c4ec661dd1e5b469919/tensorboard/pip_package/setup.py#L25-L38

File: third_party/js.bzl [hunk 14]
```diff
@@ -17,437 +17,473 @@ load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
...
-              "https://raw.githubusercontent.com/catapult-project/catapult/237aea8b58a37a2991318b6a0db60d84078e5f7e/trace_viewer_full.html"  # 2017-06-19
-          ],
-      },
-  )
+    web_library_external(
+        name = "vaadin_vaadin_split_layout",
+        licenses = ["notice"],  # Apache License 2.0
+        sha256 = "44fb83628edb77cb8392c165d4d99734750a6fbb00e5391f033962e56f14eba3",
+        urls = [
+            "https://github.com/vaadin/vaadin-split-layout/archive/v1.1.0.tar.gz",
```
Comment (stephanwlee) [inline third_party/js.bzl]: please add an entry for mirror. It should read like
`"https://mirror.bazel.build/github.com/vaadin/vaadin-split-layout/archive/v1.1.0.tar.gz"`

here and below.

File: third_party/js.bzl [hunk 15]
```diff
@@ -17,437 +17,473 @@ load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
 load("@io_bazel_rules_closure//closure:defs.bzl", "filegroup_external")
```
Comment (stephanwlee) [inline third_party/js.bzl]: Please consider removing all whitespace changes (even if it is ill-formatted) for sake of smaller diff

File: tensorboard/plugins/hparams/backend_context.py [hunk 16]
```diff
@@ -47,11 +88,222 @@ def multiplexer(self):
...
+  # This could likely be more efficiently implemented with a trie
+  # data-structure, but we don't want to add an extra dependency for that.
+  while path not in path_set:
+    if not path:
+      return None
+    path = os.path.dirname(path)
+  return path
+
+
+def _protobuf_value_type(value):
```
[the remark this answers is not in the document]
Comment (stephanwlee) [reply tensorboard/plugins/hparams/backend_context.py]: Yup! Thanks

File: tensorboard/plugins/hparams/hparams_util.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (stephanwlee) [reply tensorboard/plugins/hparams/hparams_util.py]: Nope, that's all. We need that for copybara. Thanks

File: tensorboard/plugins/hparams/hparams_demo.py [hunk 17]
```diff
@@ -113,12 +145,12 @@ def run(logdir, session_id, hparams, group_name):
     hparams: A dictionary mapping an hyperparameter name to its value.
     group_name: an id for the session group this session belongs to.
   """
-  tf.compat.v1.reset_default_graph()
-  tf.compat.v1.set_random_seed(0)
+  tf.reset_default_graph()
```
Comment (stephanwlee) [inline tensorboard/plugins/hparams/hparams_demo.py]: this file needs to be reverted to use `compat.v1`.  v2 does not have graph thus this API does not make sense.

File: tensorboard/plugins/hparams/hparams_demo.py [hunk 18]
```diff
@@ -152,30 +185,31 @@ def run(logdir, session_id, hparams, group_name):
   # coefficient. But in real life, not everything is quite so clean, so
   # we'll add in some noise. (The value of 50 is arbitrary, chosen to
   # make the data look somewhat interesting. :-) )
-  noise = 50 * tf.random.normal([])
+  noise = 50 * tf.random_normal([])
```
Comment (stephanwlee) [inline tensorboard/plugins/hparams/hparams_demo.py]: please revert the change. the API has changed to `random.normal`.

File: tensorboard/plugins/hparams/list_session_groups.py [hunk 19]
```diff
@@ -263,163 +234,385 @@ def _create_response(self, session_groups):
...
+# alternative would be to wrap the MetricName protocol buffer class in an
+# immutable class that defines equality and __hash__ methods based on the text
+# representation of the protocol buffer. This is more complex, but won't
+# require modification if we ever add fields to MetricName.
+_MetricIdentifier = collections.namedtuple('_MetricIdentifier', 'group tag')
+
+
+# We use slots here to catch typos in attributes earlier. Note that this makes
+# this class incompatible with 'pickle'.
+@attr.s(slots=True)
```
Comment (stephanwlee) [inline tensorboard/plugins/hparams/list_session_groups.py]: Is there a way to do this without the new dependency? Team seem to be against this at least for this use case.

File: tensorboard/plugins/hparams/tf_hparams_parallel_coords_plot/test/tf-hparams-parallel-coords-plot-test.html [hunk 20]
```diff
@@ -0,0 +1,310 @@
...
+          assert.deepEqual(element._pointInverseImage(scale, 0, 100), [a]);
+          assert.deepEqual(element._pointInverseImage(scale, 51, 100), []);
+          assert.deepEqual(element._pointInverseImage(scale, 50, 150), [a,b]);
+          assert.deepEqual(element._pointInverseImage(scale, 50, 249), [a,b]);
+          assert.deepEqual(element._pointInverseImage(scale, 50, 250),
+                            [a,b,c]);
+          assert.deepEqual(element._pointInverseImage(scale, 0, 300), [a,b,c]);
+        });
+
+        test("_continuousInverseImage", function() {
```
[the remark this answers is not in the document]
Comment (erzel) [reply tensorboard/plugins/hparams/tf_hparams_parallel_coords_plot/test/tf-hparams-parallel-coords-plot-test.html]: I've moved _pointInverseImage, _quantileInverseImage, _continuousInverseImage, _findClosestPath and _createScale to a public tf.hparams.parallel_coords_plot namespace.

File: tensorboard/plugins/hparams/tf_hparams_parallel_coords_plot/tf-hparams-parallel-coords-plot.html [hunk 21]
```diff
@@ -0,0 +1,837 @@
...
+
+      observers: [
+        '_redraw(sessionGroups.*)',
+        '_redraw(schema.*)',
+        '_redraw(options.*)',
+      ],
+
+      created() {
+        // To add the offsetWidth property to this.root.
+        this.style.display = "block";
```
[the remark this answers is not in the document]
Comment (erzel) [reply tensorboard/plugins/hparams/tf_hparams_parallel_coords_plot/tf-hparams-parallel-coords-plot.html]: Done. This was a copy-and-paste error.

File: tensorboard/plugins/hparams/tf_hparams_parallel_coords_plot/tf-hparams-parallel-coords-plot.html [hunk 22]
```diff
@@ -0,0 +1,837 @@
...
+      },
+
+      ready() {
+        this._container = this.$["container"];
+        this._svg = d3.select(this.$["svg"]);
+        this._sessionGroupValues = d3.select(this.$["sessionGroupValues"]);
```
[the remark this answers is not in the document]
Comment (erzel) [reply tensorboard/plugins/hparams/tf_hparams_parallel_coords_plot/tf-hparams-parallel-coords-plot.html]: Done.

File: tensorboard/plugins/hparams/tf_hparams_parallel_coords_plot/tf-hparams-parallel-coords-plot.html [hunk 23]
```diff
@@ -0,0 +1,837 @@
...
+      // Creates the DOM elements comprising the parallel-coordinate plot and
+      // registers event handlers to handle user actions such as dragging an
+      // axis and brushing.
+      // TODO(erez): Refactor this monolithic function into several objects
+      // to make the code clearer and much easier to test. We should have the
+      // following objects at least:
+      // 1. foreground lines.
+      // 2. background lines.
+      // 3. axes.
+      _draw() {
```
[the remark this answers is not in the document]
Comment (erzel) [reply tensorboard/plugins/hparams/tf_hparams_parallel_coords_plot/tf-hparams-parallel-coords-plot.html]: Ok. Yes, we should definitely make this more modular. Filed b/124243231 and b/124243365 for that.

File: third_party/js.bzl (hunk 14 above)
Comment (erzel) [reply third_party/js.bzl]: Done

File: third_party/js.bzl (hunk 15 above)
Comment (erzel) [reply third_party/js.bzl]: Done. Sorry about that.

**Comment by erzel** [comment]
Thanks @stephanwlee! I think I addressed all the comments and I have fixed the tests. PTAL.

File: third_party/js.bzl [hunk 24]
```diff
@@ -312,6 +312,45 @@ def tensorboard_js_workspace():
           "https://github.com/pair-code/facets/archive/0.2.1.tar.gz",
       ],
   )
+  web_library_external(
+      name = "vaadin_vaadin_split_layout",
+      licenses = ["notice"],  # Apache License 2.0
+      sha256 = "44fb83628edb77cb8392c165d4d99734750a6fbb00e5391f033962e56f14eba3",
+      urls = [
+          "https://github.com/vaadin/vaadin-split-layout/archive/v1.1.0.tar.gz",
+          "https://mirror.bazel.build/github.com/vaadin/vaadin-split-layout/archive/v1.1.0.tar.gz",
```
Comment (stephanwlee) [inline third_party/js.bzl]: please reorder so the mirror link comes first. Here and below. AFAIK, web_library_external will fetch in order of declaration until it HTTP 2xxs.

Comment (erzel) [reply third_party/js.bzl]: Done.

**Comment by erzel** [comment]
Thanks @stephanwlee 
Changed the mirroring order. PTAL. Note that I can't merge this myself.

 Erez.

File: tensorboard/plugins/hparams/tf_hparams_utils/test/tf-hparams-utils-test.html [hunk 25]
```diff
@@ -0,0 +1,108 @@
...
+limitations under the License.
+-->
+
+<!doctype html>
+<html>
+  <head>
+    <meta charset="utf-8">
+    <script src="../../webcomponentsjs/webcomponents-lite.min.js"></script>
+    <script src="../../web-component-tester/browser.js"></script>
+    <link rel="../../tf-imports/d3.html">
```
Comment (wchargin) [inline tensorboard/plugins/hparams/tf_hparams_utils/test/tf-hparams-utils-test.html]: Should this not be `<link rel="import" href="../../tf-imports/d3.html">`?

Why does this test work anyway?

**Comment by erzel** [comment]
Thanks. My guess is the import here is being ignored due to the syntex
error and d3 is being imported correctly from tf-hparams-utils.html which
should be <script>-included here.

 Erez.
